### PR TITLE
flatten dataclass hyperparameters for logging

### LINKED ISSUE #7568: TensorBoardLogger does not store hparams inside a dataclass
## Please reproduce using the BoringModel

When hyper-parameters are wrapped by a `dataclass` and passed to a LightningModule, the `TensorBoardLogger` will store a string representation of the entire `dataclass` as one hyper-parameter instead of the contents inside the `dataclass`:

tensorboard

### To Reproduce

Use following [notebook](https://colab.research.google.com/drive/1DFqkyzt-qL5yyZ_IWDpcZV3BGZNVFzWA?usp=sharing)

### Expected behavior

The hyper-params should be unwrapped, as is done for dictionaries, such that the HPARAMS tab in Tensorboard shows individual hyper-params, e.g.:
- `model/n_inputs`
- `model/n_outputs`
- `training/learning_rate`

Since a `dataclass` can be converted to a dict via `datalcasses.asdict`, it shouldn't be too hard to add a another case to 
[LightningLoggerBase._flatten_dict](https://github.com/PyTorchLightning/pytorch-lightning/blob/e126649d19e85c449b007008361f10374878f2f4/pytorch_lightning/loggers/base.py#L240)

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu101
	- pytorch-lightning: 1.3.1
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Tue Apr 20 19:55:43 PDT 2021

### PR DESCRIPTION
**Pull request by jaswon**
## What does this PR do?

Fixes #7568

Flattens dataclass objects (in a similar fashion to dicts) for the purpose of logging hyperparameters

<details>
  <summary><b>Before submitting</b></summary>

- [x] Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

----
:books: Documentation preview :books:: https://pytorch-lightning--18906.org.readthedocs.build/en/18906/

### CODE REVIEW

File: src/lightning/fabric/utilities/logger.py [hunk 1]
```diff
@@ -88,8 +89,11 @@ def _flatten_dict(params: MutableMapping[Any, Any], delimiter: str = "/", parent
     result: Dict[str, Any] = {}
     for k, v in params.items():
         new_key = parent_key + delimiter + str(k) if parent_key else str(k)
-        if isinstance(v, Namespace):
+        if is_dataclass(v):
+            v = asdict(v)
```
Comment (awaelchli) [inline src/lightning/fabric/utilities/logger.py:93]: Do you know, does a dataclass match the type `MutableMapping[Any, Any]` as annotated in the function?

File: tests/tests_fabric/utilities/test_logger.py [hunk 2]
```diff
@@ -73,6 +74,26 @@ def test_flatten_dict():
...
+        b: int
+
+    params = {"params": B(a=A(c=1, d=2), b=3), "param": 4}
+    params = _flatten_dict(params)
+    assert type(params) is dict
+    assert params.pop("param") == 4
+    assert params.pop("params/b") == 3
+    assert params.pop("params/a/c") == 1
+    assert params.pop("params/a/d") == 2
+    assert len(params) == 0
```
Comment (awaelchli) [inline tests/tests_fabric/utilities/test_logger.py:95]: Could you `assert params == the contents` directly?

**Review by awaelchli — APPROVED** [review]
Looks reasonable :) Thanks!

File: src/lightning/fabric/utilities/logger.py (hunk 1 above)
Comment (jaswon) [reply src/lightning/fabric/utilities/logger.py:93]: the type of `params` is `MutableMapping[Any, Any]`, thus the type of `v` is `Any`, so a dataclass does match the type

**Review by carmocca — APPROVED** [review]
